### ISSUE TITLE
Research: Redesign of the Methods

### ISSUE THREAD

**Issue by HallM** [Crazy50/jsframework#31]
Actions/Methods whatever to call them

_ALL_ actions originate from a client. be it over http, ws, rpc, whatever. it comes from a client.

there may be a desire for a client to send actions to other clients though. server filters/controls/routes
but this is just much later down the road. has so many potential issues to deal with like tracking clients

types of Actions:
- Universally executable (doesn't matter who calls it)
  - must have either:
    - a single "handler"
      -- actually... is there really ever a case for a single handler? it's just a plain function at that point.
      -- like seriously. theres not really "universal-state", therefore what could you manipulate? and why not just use a plain function?
      -- MAYBE something that does multiple things at the same time, but only want to make 1 server request?
    - a "server" and a "client" separate handlers
  - server portion is executed via ajax or form-post
    - post is a post-redirect type of situation
- Server-side executable (client side must ajax it over)
  - has a server side handler
  - may have a "try optimistic update" client handler after sending, but before return
    - though, optimistic is _difficult_ for sure
  - may have a client handler after server returns
  - in some ways, this is a universal handler with user-defined "server" and pre-built "client"
- Client-side executable (the client handles action by itself, the server never knows)
  - cant work without JS, but alas #DealWithIt I guess

execution really is up to the user
they could execute in a universal friendly manner or helper which will call handler/client when possible
or make a POST-redirect style request to the server if not.

maybe they use a Form or Link helper that sets the action/href plus does "client"-handler when possible (universal)
maybe they use form or anchor and set the action/href themselves plus attach the "client"-handler (universal still)
maybe they use whatever else, dont care about the action/href, but attach the "client"-handler (client only)
maybe they call it as a function within the client-side code
maybe they use a form or anchor and set the action/href, but don't attach the "client"-handler (server-side)

doesn't really make sense to call it as a function within the server-side code...
basically, should methods ever be called from another method? that's really the only case.
without that, the point for a "single handler" are impossible. so it may be desirable.
example: doThis() { iDidIt=true; doSomethingElse(); } where doSomethingElse is also a method that can be called alone

external APIs? - maybe just universal executable and they make the call themselves
modifying server state? uhhh... modify a DB or session or just whatever
modifying client state? well... uh. guess we need some sort of state system. crap.
is there even such a thing as a "universal" action that runs exactly the same code client and server side?
  cant really think of a case. there's not really universal-state. at that point, all you're left with is a function.

**Comment by HallM** [comment]
I think this covers all cases:

```
-- Server only, not even callable outside or client. but hey, design by contract maybe.
Method({
  params: {},
  server: function(params) {
    return data || promise;
  },
  client: false
})

-- server/restful only. cause you might be building just an api. why not.
Method({
  url: '/path/to/service', // not dynamic for reasons at this time
  method: 'post',
  params: {},
  server: function(params) {
    return data || promise;
  },
  client: false
})

-- server-side, client auto-create rest-call
Method({
  url: '/path/to/service', // actually required for once, because we need to know where we call it
  method: 'post',
  params: {},
  server: function(params) {
    return data || promise;
  },
  client: true // just auto generate whatever
})

-- server-side, client auto-create rest-call with pre/post/error delegates
Method({
  url: '/path/to/service', // actually required for once, because we need to know where we call it
  method: 'post',
  params: {},
  server: function(params) {
    return data || promise;
  },
  client: {
    pre: function(params) {}, // someone might do it, do optimistic type stuff here?
    post: function(result) {}, // after the server returns stuff
    error: function(error) {} // do something with the error
  }
})

-- server-side, client separate handlers
Method({
  // maybe you could skip url+method. maybe client can handle it by itself and doesnt need server
  params: {},
  server: function(params) {
    // does one thing if called via server
    return data || promise;
  },
  client: function(params) {
    // can do whatever you want on the client side
  }
})

-- client side only
Method({
  params: {},
  client: function(params) {
    // do the stuff
  }
})

-- uhhhh.... well. someone could really like the design by contract.
Method({
  params: {},
  both: function(params) {
    // do the stuff
  }
})

-- why not
Method({
  url: '/thing',
  method: 'post',
  params: {},
  both: function(params) {
    // do the stuff
  }
})
```

Cases I'm completely dropping from previous iterations:
- automatic creation of calls to external APIs. just let them do it themselves. Just expose Axios (or let them use anything else)
